### PR TITLE
DA gap fills

### PR DESCRIPTION
**Pull request by jhrehanoaa**
This PR should accurately calculate the DA gaps based on NCAR's calculation.

 FVD = (Last_obs-Troute) * decay_weight + last_obs

I have made a few changes to structured and structured-obj in order to complete this capability. 

The main code changes involve tracking an array of the current decay timesteps and the current last observations. Whenever a non-NaN value is discovered in the USGS array it replaces the current last obs and resets the decay timestep. If a NaN value is found it reaches back into the gage_i location of both the decay timestep array and last obs array in order to calculate the new delta, decay_weight, and ultimately the new FVD replacement. Values should match at the gage and be slightly different above and below. I am not sure if we would like to try to implement the decay function after the USGS observations end as it would be changing all predictions we make using route essentially. 

        if gages_size:

            for gage_i in range(gages_size):                        
                usgs_position_i = usgs_positions_list[gage_i]
                a = 120  # TODO: pull this a value from the config file somehow  
                for i in range(0,len(usgs_values[0])):
                    if np.isnan(usgs_values[gage_i, timestep-1]):
                        da_weight = exp(decay_timesteps[gage_i]/-a)
                        flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                        #alternative version below using decay_timestep not da_weight calculation based on discussions 
                        #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                        decay_timesteps[gage_i] += 1
                    else:
                        decay_timesteps[gage_i] = 1
                        starting_bias_array[gage_i] = usgs_values[gage_i, timestep-1]
                        flowveldepth[usgs_position_i, timestep , 0] = usgs_values[gage_i, timestep-1]
                        pass

### CODE REVIEW

**Comment by jhrehanoaa** [comment]
python3 -m nwm_routing -V2 -f "../../test/input/yaml/Florence_Benchmark_da.yaml"

command needed to run.

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx [hunk 1]
```diff
@@ -349,103 +353,117 @@ cpdef object compute_network(
     # minimal validation,
     # Jump straight to nogil.
 
-    with nogil:
```
Comment (jameshalgren) [inline src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:352]: We have to support the nogil

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx [hunk 2]
```diff
@@ -902,76 +937,82 @@ cpdef object compute_network_structured_obj(
...
-        previous_upstream_flows = 0.0
-        for id in r.upstream_ids: #Explicit loop reduces some overhead
-          upstream_flows += flowveldepth[id, timestep, 0]
-          previous_upstream_flows += flowveldepth[id, timestep-1, 0]
+            #Need to get quc and qup
+            upstream_flows = 0.0
+            previous_upstream_flows = 0.0
+            for id in r.upstream_ids: #Explicit loop reduces some overhead
+                upstream_flows += flowveldepth[id, timestep, 0]
+                previous_upstream_flows += flowveldepth[id, timestep-1, 0]
```
Comment (jameshalgren) [inline src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:947]: Will you confirm that this is exclusively a formatting update?

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx [hunk 3]
```diff
@@ -902,76 +937,82 @@ cpdef object compute_network_structured_obj(
...
+                    flowveldepth[id, timestep, 2] = out_buf[i, 2]
+           
+        if gages_size:
+
+            for gage_i in range(gages_size):                        
+                usgs_position_i = usgs_positions_list[gage_i]
+                a = 120  # TODO: pull this a value from the config file somehow  
+                for i in range(0,len(usgs_values[0])):
+                    if np.isnan(usgs_values[gage_i, timestep-1]):
+                        da_weight = exp(decay_timesteps[gage_i]/-a)
```
Comment (jameshalgren) [inline src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:1003]: This decay coefficient needs to be compared in similar units -- a is 120 minutes, so the `decay_timestep[gage_i]` needs to be converted to a time in minutes (presumably by multiplying by dt).

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx [hunk 4]
```diff
@@ -1265,11 +1312,42 @@ cpdef object compute_network_structured(
...
+
+            for gage_i in range(gages_size):                        
                 usgs_position_i = usgs_positions_list[gage_i]
-                if timestep < gage_maxtimestep:
-                    flowveldepth[usgs_position_i, timestep, 0] = usgs_values[gage_i, timestep-1]
-               
+                a = 120  # TODO: pull this a value from the config file somehow  
+                for i in range(0,len(usgs_values[0])):
+                    if np.isnan(usgs_values[gage_i, timestep-1]):
+                        da_weight = exp(decay_timesteps[gage_i]/-a)
```
Comment (jameshalgren) [inline src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:1321]: See comment above about the ratio of the timestep with the decay coefficient.

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx [hunk 5]
```diff
@@ -1200,8 +1247,8 @@ cpdef object compute_network_structured(
     cdef float reservoir_outflow, reservoir_water_elevation
     cdef int id = 0
     #Run time
-    with nogil:
-      while timestep < nsteps+1:
+    #with nogil:
```
Comment (jameshalgren) [inline src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:1250]: We have to support the nogil statement.

File: test/input/yaml/Croton_NY_hybrid_usgs.yaml [hunk 6]
```diff
@@ -1,14 +1,14 @@
...
+    parallel_compute_method: by-subnetwork-jit  # OPTIONS: <omit flag for serial execution>, "by-network", "by-subnetwork-jit", "by-subnetwork-jit-clustered"
     # subnetwork_target_size: 100  # by-subnetwork* requires a value here to identify the target subnetwork size.
     verbose: true  # verbose output (leave blank for quiet output.)
     showtiming: true  # set the showtiming (omit flag for no timing information.)
     debuglevel: 1  # set the debuglevel for additional console output.
     break_network_at_waterbodies: true # replace waterbodies in the route-link dataset with segments representing the reservoir and calculate to divide the computation (leave blank for no splitting.)
                           # WARNING: `break_network_at_waterbodies: true` will only work if compute_method is set to "V02-structured-obj" and parallel_compute_method is unset (serial execution) or set to "by-network".
-    compute_method: V02-structured-obj  # OPTIONS: "V02-caching", "V02-structured-obj", "V02-structured"
+    compute_method: V02-structured  # OPTIONS: "V02-caching", "V02-structured-obj", "V02-structured"
```
Comment (jameshalgren) [inline test/input/yaml/Croton_NY_hybrid_usgs.yaml:11]: Let's omit these changes unrelated to the PR

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx (hunk 5 above)
Comment (jhrehanoaa) [reply src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:1250]: #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                            starting_bias_array[gage_i] = usgs_values[gage_i, timestep-1]
                                                                    ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1328:69: Converting to Python object not allowed without gil


I am getting a lot of these errors when trying to run with nogil. Is there a simple workaround for cython? The assigning has to happen.

Comment (jhrehanoaa) [reply src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:1250]: A large number of errors occurring once Gil is shut off. 

Error compiling Cython file:
------------------------------------------------------------
...
                        
            if gages_size:

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                   ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1318:20: Assignment of Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1327:43: Assignment of Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                            starting_bias_array[gage_i] = usgs_values[gage_i, timestep-1]
                                              ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1328:47: Assignment of Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:35: Truth-testing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:35: Calling gil-requiring function not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:29: Accessing Python attribute not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:27: Accessing Python global or builtin not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:35: Constructing Python tuple not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                                              ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:47: Converting to Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                                                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1321:67: Coercion from Python not allowed without the GIL

Error compiling Cython file:
------------------------------------------------------------
...
                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                                                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1321:67: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1321:59: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                                                                   ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1321:68: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:163: Coercion from Python not allowed without the GIL

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:163: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                      ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:151: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                      ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:103: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                             ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:94: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                    ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:117: Converting to Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                        ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:153: Converting to Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                                                       ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:184: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1325:43: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1325:43: Calling gil-requiring function not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1325:43: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1327:43: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                            starting_bias_array[gage_i] = usgs_values[gage_i, timestep-1]
                                              ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1328:47: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                            starting_bias_array[gage_i] = usgs_values[gage_i, timestep-1]
                                                                    ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1328:69: Converting to Python object not allowed wi

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx (hunk 2 above)
Comment (jhrehanoaa) [reply src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:947]: Should be. We moved some indentations that were half indented around in some places to make it look cleaner. Still works.

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx (hunk 1 above)
Comment (jhrehanoaa) [reply src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:352]: I tried adding cdef comments for all available variables that are erroring. However, I am still receiving plenty of python conversion, Coercion, and indexing errors.

File: src/nwm_routing/src/nwm_routing/__main__.py [hunk 7]
```diff
@@ -486,17 +494,30 @@ def main_v02(argv):
...
+
+    last_obs_df = pd.DataFrame()
+
+    if last_obs_file:
+        usgs_df, last_obs_df, last_obs_date = nnu.build_data_assimilation(
+            data_assimilation_parameters
+        )
+        last_obs_start = nhd_io.get_last_obs_location(qlats, last_obs_date)
+
+    else:
```
Comment (jameshalgren) [inline src/nwm_routing/src/nwm_routing/__main__.py:511]: Are there corner cases for this logic? (i.e., `data_assimilation_filter == true` and `last_obs_file == false` with some condition that requires a 'last_obs` to jumpstart the data assimilation....)

File: src/nwm_routing/src/nwm_routing/__main__.py [hunk 8]
```diff
@@ -486,17 +494,30 @@ def main_v02(argv):
...
+        last_obs_start = -1
+
+    if data_assimilation_csv or data_assimilation_folder or data_assimilation_filter:
         if showtiming:
             start_time = time.time()
         if verbose:
             print("creating usgs time_slice data array ...")
 
-        usgs_df, _ = nnu.build_data_assimilation(data_assimilation_parameters)
-
```
Comment (jameshalgren) [inline src/nwm_routing/src/nwm_routing/__main__.py:499]: Without this line in here, this logic has no effect... was it intentionally omitted?

File: src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx [hunk 9]
```diff
@@ -177,6 +177,7 @@ cpdef object compute_network(
     const float[:,:] usgs_values,
     const int[:] usgs_positions_list,
     const float[:,:] lastobs_values,
+    int last_obs_start,
```
Comment (jameshalgren) [inline src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx:180]: This function header is not compatible with the other calls ...

**Comment by jameshalgren** [comment]
Pulling the key technical comment and putting it into #358, which will replace this PR.

**Pull request closed without merging**